updated documentation for fallback credentials (#91)
CiscoTestAutomation/unicon.plugins: docs/user_guide/passwords.rst

Password Handling
=================

Passwords are defined in the testbed YAML file. This document describes the
password handling logic used by the different plugins.

> For understanding the password handling first we need to understand credentials on pyATS.

Reviewer: ```suggestion
Passwords are managed through device credentials in pyATS, the next section explains how credentials are handled in pyATS.
```

.. _credentails:

Credentials
-----------

The ``credentials`` connection parameter defines a dictionary of named
credentials.  A credential is a dictionary typically containing both
``username`` and ``password`` keys.

The ``login_creds`` connection parameter defines an optional sequence of
credential names to try.  Each time the device prompts for a username or
password, the current credential is set to the next credential in the sequence
if a current credential has not already been set.
When a password is sent, the current credential is unset.  The one exception
is when entering an administrator's password on a routing device coming up
without configuration, in this case the current credential is reused.
If the sequence has been exhausted and no more credentials are available to
satisfy a username/password prompt, a
`CredentialsExhaustedError<unicon.core.errors.CredentialsExhaustedError>` is
raised.

> Credentials are not retried, any username or password failure causes a
> `UniconAuthenticationError<unicon.core.errors.UniconAuthenticationError>`
> to be raised.

Reviewer: ```suggestion
Credentials are not retried, any username or password failure causes a
`UniconAuthenticationError<unicon.core.errors.UniconAuthenticationError>`
to be raised, unless fallback credentials are defined. See :ref:`fallback_credentials`.
```

``login_creds`` is used to describe the order of credentials to use on
initial login.  If not specified, the ``default`` credential is used.

> Please see :ref:`credentails` for more details.

Reviewer: ```suggestion
Please see :ref:`credentials` for more details.
```

    bad_passwords = r'^.*?% (Bad passwords|Access denied|Authentication failed)'

> 

Reviewer: ```suggestion

.. _fallback_credentails:

```

Fallback Credentials
--------------------

> In case of authentication failure,
>  you could use fallback credentials before erroring out.
> you could have a couple of login credentials and define them using fallback credentials.
> These login credentials will be used in sequence. If none of the combination works on the device
> we get the bad password exception.
> 
> you could have a default list for all the connections in the testbed:

Reviewer: ```suggestion
In case of authentication failure, fallback credentials are used to try and login to the device when they are defined for the connection.  You can define one or more device credentials and define them as fallback credentials by adding them to the fallback credentials list under the `defaults` section or under the connection. The fallback credentials credentials will be used in sequence. If none of the combinations work on the device a bad password exception is raised.

Below an example of a testbed with fallback credentials defined.
```

.. code-block:: yaml

    devices:
      my_device:
        type: router
        os: ios
        credentials:
            default:
                username: cisco
                password: secret
            enable:
                password: enable
            set_1:
              password: lab
              username: lab
            set_2:
              password: admin
              username: admin
        connections:
            defaults:
              class: unicon.Unicon
              fallback_credentials:
                - set1
                - set2
            netconf:
              class: yang.connector.Netconf
              ip: 1.2.3.4
              port: 830
              protocol: netconf
            telnet:
              ip: 1.2.3.4
              port: 23
              protocol: telnet

> or you could define fallback credentails per connection:

Reviewer: ```suggestion
Example of fallback credentials per connection:
```

                - set1
                - set2
            netconf:
              class: yang.connector.Netconf
              ip: 1.2.3.4
              port: 830
              protocol: netconf
            telnet:
              ip: 1.2.3.4
              port: 23
              protocol: telnet

Example of fallback credentials per connection:

.. code-block:: yaml

    devices:
      my_device:
        type: router
        os: ios
        credentials:
            default:
                username: cisco
                password: secret
            enable:
                password: enable
        connections:
            netconf:
              class: yang.connector.Netconf
              ip: 1.2.3.4
              port: 830
              protocol: netconf
              fallback_credentials:
                  - set_1
              credentials:
                  default:
                      username: cisco
                      password: secret
                  set_1:
                    password: lab
                    username: lab
            telnet:
              ip: 1.2.3.4
              port: 23
              protocol: telnet
              fallback_credentials:

>                 - set_2

Reviewer: set_2 is not defined

```suggestion
                - set_1
```
